// webdriver/build/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULTS = void 0;
const DEFAULTS = {
  protocol: {
    type: 'string',
    default: 'http',
    match: /(http|https)/
  },
  hostname: {
    type: 'string',
    default: 'localhost'
  },
  port: {
    type: 'number',
    default: 4444
  },
  path: {
    type: path => {
      if (typeof path !== 'string') {
        throw new TypeError('The option "path" needs to be from type "string"');
      }

      if (!path.startsWith('/')) {
        throw new TypeError('The option "path" needs to start with a "/"');
      }

      return true;
    },
    default: '/wd/hub'
  },
  queryParams: {
    type: 'object'
  },
  capabilities: {
    type: 'object',
    required: true
  },
  logLevel: {
    type: 'string',
    default: 'info',
    match: /(trace|debug|info|warn|error|silent)/
  },
  connectionRetryTimeout: {
    type: 'number',
    default: 90000
  },
  connectionRetryCount: {
    type: 'number',
    default: 3
  },
  user: {
    type: 'string'
  },
  key: {
    type: 'string'
  },
  agent: {
    type: 'object'
  },
  headers: {
    type: 'object'
  },
  enableDirectConnect: {
    type: 'boolean',
    default: false
  },
  transformRequest: requestOptions => requestOptions,
  transformResponse: response => response
};
exports.DEFAULTS = DEFAULTS;
